### PR TITLE
Block backward incompatible schema change

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Currently when uploading a backward incompatible schema, we only log a warning instead of blocking the schema update.
This PR changes the behavior to block the backward incompatible schema change.

## Release Notes
Backward incompatible schema change through controller rest API `PUT /schemas/{schemaName}` will be blocked.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java [hunk 1]
```diff
@@ -148,9 +150,10 @@ public SuccessResponse updateSchema(
   @ApiResponses(value = {@ApiResponse(code = 200, message = "Successfully updated schema"), @ApiResponse(code = 404, message = "Schema not found"), @ApiResponse(code = 400, message = "Missing or invalid request body"), @ApiResponse(code = 500, message = "Internal error")})
   public SuccessResponse updateSchema(
       @ApiParam(value = "Name of the schema", required = true) @PathParam("schemaName") String schemaName,
+      @ApiParam(value = "Whether to force update the schema if the new schema is not backward compatible") @DefaultValue("false") @QueryParam("force") boolean force,
       @ApiParam(value = "Whether to reload the table if the new schema is backward compatible") @DefaultValue("false") @QueryParam("reload") boolean reload,
       Schema schema) {
-    return updateSchema(schemaName, schema, reload);
+    return updateSchema(schemaName, schema, force, reload);
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:156]: Shall we not have `force` parameter?
When do we really expect user to force a backward incompatible change? They might interpret this as "by default backward incompatible changes are not allowed but it can be allowed if this is set to true"

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 2]
```diff
@@ -1066,14 +1067,22 @@ public void updateSchema(Schema schema, boolean reload)
...
+
     PinotHelixPropertyStoreZnRecordProvider propertyStoreHelper =
         PinotHelixPropertyStoreZnRecordProvider.forSchema(_propertyStore);
     propertyStoreHelper.set(schemaName, record);
 
-    boolean isNewSchemaBackwardCompatible = schema.isBackwardCompatibleWith(oldSchema);
-    if (!isNewSchemaBackwardCompatible) {
-      LOGGER.warn(String.format("New schema %s is not backward compatible", schemaName));
-    } else if (reload) {
+    if (isNewSchemaBackwardCompatible && reload) {
```
Comment (siddharthteotia) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1085]: Eventually user will have to reload for schema change to take effect on pre-existing segments and it will fail.

One viable scenario for forcing a backward incompatible change could be for REFRESH use cases. For example, changing a column type from LONG to INT followed by starting a refresh.

For APPEND use cases, we recently had a situation where 2 years ago the column type was changed from LONG to INT and segments were never reloaded. So, the user was under the impression that schema change went through for old segments as well. Their recent selection queries somehow ended up touching old segments and failed the schema consistency check in SelectionOnlyCombineOperator. So, for APPEND use cases may be we should always disallow ?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java [hunk 3]
```diff
@@ -243,39 +246,38 @@ private SuccessResponse addSchema(Schema schema, boolean override) {
       return new SuccessResponse(schemaName + " successfully added");
     } catch (Exception e) {
       _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SCHEMA_UPLOAD_ERROR, 1L);
-      throw new ControllerApplicationException(LOGGER,
-          String.format("Failed to add new schema %s.", schemaName), Response.Status.INTERNAL_SERVER_ERROR,
-          e);
+      throw new ControllerApplicationException(LOGGER, String.format("Failed to add new schema %s.", schemaName),
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:249]: Won't this end up throwing 5xx if schema update failed due to backward incompatibility?

Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:249]: This is auto-formatting change under the `addSchema()` function. We will throw 400 (BAD_REQUEST) error when users making backward incompatible schema update (line 292).

**Comment by Jackie-Jiang** [comment]
@siddharthteotia @mcvsubbu I can remove the force update flag. Then in order to make schema incompatible change, user has to drop and recreate the table. wdyt?

**Comment by mcvsubbu** [comment]
> @siddharthteotia @mcvsubbu I can remove the force update flag. Then in order to make schema incompatible change, user has to drop and recreate the table. wdyt?

That is totally fine I think. The way I look at this is as follows:

 If we have a production deployment, it is good  in multiple respects to NOT have backward incompatible changes. New columns can be added, and old columns just set to default value in future segments. In a columnar database, the only extra resource consumed is metadata for the column.

If this is a non-production env (experimental) it should be ok to drop the data and recreate the table. Maybe we can provide some recipes for saving off existing segments and recreating the segments for the new table. The main problem that users may have is if they drop the table and recreate it the same instant (before the old table goes away). We could (should) add a check in table creation logic that there is no EXTERNALVIEW for a new table being created. If there is, fail the table with a good error message to ask the user to try later.

Bottom line: Good to drop the table in either case.

**Comment by Jackie-Jiang** [comment]
@siddharthteotia @mcvsubbu Removed the force flag as suggested. Please take another look

**Review by mcvsubbu — APPROVED** [review]
